easycv/apis/export.py

```diff
@@ -372,3 +445,199 @@ def replace_syncbn(backbone_cfg):
+            }
+
+            return out
+else:
+    PreProcess = None
+    DetPostProcess = None
+
+
+class End2endModelExportWrapper(torch.nn.Module):
+    """Model export wrapper that supports end-to-end export of pre-processing and post-processing.
+    We support some built-in preprocessing and postprocessing functions.
+    If the requirements are not met, you can customize the preprocessing and postprocessing functions.
+    The custom functions must support satisfy requirements of `torch.jit.script`,
+    please refer to: https://pytorch.org/docs/stable/jit_language_reference_v2.html
+
+    Args:
+        model (torch.nn.Module):  `torch.nn.Module` that will be run with `example_inputs`.
+            `model` arguments and return values must be tensors or (possibly nested) tuples
+            that contain tensors. When a module is passed `torch.jit.trace`, only the
+            ``forward_export`` method is run and traced (see :func:`torch.jit.trace
+            <torch.jit.trace_module>` for details).
+        example_inputs (tuple or torch.Tensor):  A tuple of example inputs that
+            will be passed to the function while tracing. The resulting trace
+            can be run with inputs of different types and shapes assuming the
+            traced operations support those types and shapes. `example_inputs`
+            may also be a single Tensor in which case it is automatically
+            wrapped in a tuple.
+        preprocess_fn (callable or None): A Python function for processing example_input.
+            If there is only one return value, it will be passed to `model.forward_export`.
+            If there are multiple return values, the first return value will be passed to `model.forward_export`,
+            and the remaining return values ​​will be passed to `postprocess_fn`.
+        postprocess_fn (callable or None): A Python function for processing the output value of the model.
+            If `preprocess_fn` has multiple outputs, the output value of `preprocess_fn`
+            will also be passed to `postprocess_fn`. For details, please refer to: `preprocess_fn`.
+        trace_model (bool): If True, before exporting the end-to-end model,
+            `torch.jit.trace` will be used to export the `model` first.
+            Traceing an ``nn.Module`` by default will compile the ``forward_export`` method and recursively.
+
+    Examples:
+        import torch
+
+        batch_size = 1
+        example_inputs = 255 * torch.rand((batch_size, 3, 640, 640), device='cuda')
+        end2end_model = End2endModelExportWrapper(
+            model,
+            example_inputs,
+            preprocess_fn=PreProcess(target_size=(640, 640)),  # `PreProcess` refer to ev_torch.apis.export.PreProcess
+            postprocess_fn=DetPostProcess()  # `DetPostProcess` refer to ev_torch.apis.export.DetPostProcess
+            trace_model=True)
+
+        model_script = torch.jit.script(end2end_model)
```

Comment:
Could we use `torch.jit.trace` mechanism to export an End2End model here, or do you think the inference time will be slower if the script mechanism is used here?

Reply:
We support to export an End2End model with both the preprocess and postprocess procedure by torch.jit.script, which can not be replaced by the torch.jit.trace since the input must be a tensor. You can also export a model with ''end2end=False'' that use trace to export a model and it is little bit faster.